src/main/java/life/qbic/openbis/openbisclient/OpenBisClient.java

```diff
@@ -904,10 +978,164 @@ public URL getDataStoreDownloadURLLessGeneric(String dataSetCode, String openbis
    *
    * @param projectCode The 5 letter QBiC code of the project
    * @param sampleType The openBIS sampleType that should be included in the result
+   * @return List containing each line of the resulting TSV fil
    */
-  @Override
   public List<String> getProjectTSV(String projectCode, String sampleType) {
-    return null;
+    List<String> res = new ArrayList<String>();
+    // search all samples of project
+    List<Sample> allSamples = getSamplesOfProject(projectCode);
+    // filter all samples by types
+    List<Sample> samples = new ArrayList<Sample>();
+    for (Sample s : allSamples) {
+      if (sampleType.equals(s.getType().getCode()))
+        samples.add(s);
+    }
+    // sort remaining samples-
+    // Arrays.sort(samples);
+
+    Map<String, String> ncbi = new HashMap<>();
+    Map<String, String> taxonomyMap = getVocabCodesAndLabelsForVocab("Q_NCBI_TAXONOMY");
+    for (String key : taxonomyMap.keySet()) {
+      ncbi.put(taxonomyMap.get(key), key);
+    }
+    String header = "QBiC Code\tSecondary Name\tLab ID\tSample Type\tAttributes\tSource";
+    if (!sampleType.equals("Q_BIOLOGICAL_ENTITY"))
+      header += "\tSource Name(s)\tSource Lab ID(s)";
+    if (sampleType.equals("Q_TEST_SAMPLE"))
+      header += "\tExtract Code(s)\tExtract Name(s)\tExtract Lab ID(s)";
+    res.add(header);
+    for (Sample sample : samples) {
+      String code = sample.getCode();
+      List<String> row = new ArrayList<String>();
+      row.add(code);
+      String secName = sample.getProperties().get("Q_SECONDARY_NAME");
+      if (secName == null)
+        secName = "";
+      row.add(secName);
+      String extID = sample.getProperties().get("Q_EXTERNALDB_ID");
+      if (extID == null)
+        extID = "";
+      row.add(extID);
+      String extrType = sample.getProperties().get("Q_PRIMARY_TISSUE");
+      if (extrType == null)
+        extrType = sample.getProperties().get("Q_SAMPLE_TYPE");
+      if (extrType == null)
+        extrType = "";
+      if (extrType.equals("CELL_LINE"))
+        extrType = sample.getProperties().get("Q_TISSUE_DETAILED");
+      row.add(extrType);
+      String props = sample.getProperties().get("Q_PROPERTIES");
+      if (props == null)
```

Comment:
Suggestion: These null checks are quite cluttering the code and decrease readability. What a about a small helper method?

```
private static String getSampleProperty(Sample sample, String propertyType) {
    String property = sample.getProperty().get(propertyType)
    String result = (property == null) ? "" : property;
    return property;
} 
```

Reply:
changed